src/main/java/xyz/nucleoid/plasmid/util/BucketScanner.java

```diff
@@ -0,0 +1,253 @@
+package xyz.nucleoid.plasmid.util;
+
+import it.unimi.dsi.fastutil.longs.LongArraySet;
+import it.unimi.dsi.fastutil.longs.LongSet;
+import net.minecraft.block.Block;
+import net.minecraft.server.world.ServerWorld;
+import net.minecraft.structure.rule.RuleTest;
+import net.minecraft.tag.Tag;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+
+import java.util.*;
+import java.util.function.BiPredicate;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Methods for finding connected blocks as a {@link LongSet}. Use {@link BlockPos#fromLong} to retrieve returned positions.
+ */
+public final class BucketScanner {
+    private final Builder builder;
+
+    private final Map<Long, Integer> depthMap;
+
+    private final Deque<Long> scanQueue;
+    private final BlockPos.Mutable previousMutable;
+    private final BlockPos.Mutable nextMutable;
+
+    private BucketScanner(Builder builder, BlockPos origin) {
+        this.builder = builder;
+        this.depthMap = new HashMap<>();
+
+        this.scanQueue = new ArrayDeque<>();
+        this.previousMutable = origin.mutableCopy();
+        this.nextMutable = origin.mutableCopy();
+
+        scanQueue.push(origin.asLong());
+        depthMap.put(origin.asLong(), builder.maxDepth);
+
+        this.initialize();
+    }
+
+    /**
+     * Creates a bucket scanner builder.
+     *
+     * @param predicate the predicate for the blocks that can be accepted in the set
+     * @return a new bucket scanner builder
+     */
+    public static Builder create(ScanPredicate predicate) {
+        return new Builder(predicate, Connectivity.SIX, SearchType.DEPTH_FIRST, 512, -1);
+    }
+
+    /**
+     * Creates a bucket scanner builder.
+     *
+     * @param block the block type that can be accepted in the set
+     * @return a new bucket scanner builder
+     */
+    public static Builder create(Block block, ServerWorld world) {
```

Comment:
I would recommend adding an additional method that has a `BlockState` parameter.